### PR TITLE
`arrow2-convert` migration 3: `attr.rust.override_crate` & optimizing out dependencies

### PR DESCRIPTION
**Pull request by teh-cmc**
This introduces the `rust.attr.override_crate` attribute, which is necessary for the serde-codegen story.

This attribute also allows us to generate some fundamental types such as `InstanceKey` directly into `re_types_core` rather than `re_types`, making it possible to reduce the number of crates that depend on the giant `re_types`.

- The `AutoSpaceViews` & `PanelView` components are now back into their respective home (`re_viewport` & `re_viewer`, respectively).
They were temporarily moved it because we had no support for `custom_crate`.
They will be joined by their more complicated siblings in the next PR, which implements the necessary serde-codegen support.

- `InstanceKey`, `Clear` and `ClearIsRecursive` are now generated into `re_types_core`.
This allows `re_query`, `re_arrow_store` and `re_data_store` to drop their dependency on `re_types`.

- Generated code now picks up arrow from `re_types_core::external` instead of importing it directly.
This means crates hosting generated code are not forced into import `arrow2` directly.

---

`arrow2-convert` migration PR series:
- #3853 
- #3855
- #3897
- #3902 

---

### Checklist
* [x] I have read and agree to [Contributor Guide](https://github.com/rerun-io/rerun/blob/main/CONTRIBUTING.md) and the [Code of Conduct](https://github.com/rerun-io/rerun/blob/main/CODE_OF_CONDUCT.md)
* [x] I've included a screenshot or gif (if applicable)
* [x] I have tested [demo.rerun.io](https://demo.rerun.io/pr/3897) (if applicable)
* [x] The PR title and labels are set such as to maximize their usefulness for the next release's CHANGELOG

- [PR Build Summary](https://build.rerun.io/pr/3897)
- [Docs preview](https://rerun.io/preview/11654c9c77cb1e7feebe637f3234e4a041088539/docs) 
- [Examples preview](https://rerun.io/preview/11654c9c77cb1e7feebe637f3234e4a041088539/examples) 
- [Recent benchmark results](https://ref.rerun.io/dev/bench/)
- [Wasm size tracking](https://ref.rerun.io/dev/sizes/)

### CODE REVIEW

File: crates/re_types/definitions/rerun/blueprint/auto_space_views.fbs [hunk 1]
```diff
@@ -15,8 +15,10 @@ namespace rerun.blueprint;
 struct AutoSpaceViews (
   "attr.arrow.transparent",
   "attr.rust.derive": "Copy, Default",
+  "attr.rust.override_crate": "re_viewport",
   "attr.rust.repr": "transparent",
   "attr.rust.tuple_struct"
 ) {
   enabled: bool (order: 100);
 }
+
```
Comment (teh-cmc) [inline crates/re_types/definitions/rerun/blueprint/auto_space_views.fbs:24]: x

File: crates/re_types/definitions/rerun/blueprint/panel_view.fbs [hunk 2]
```diff
@@ -13,7 +13,9 @@ namespace rerun.blueprint;
 ///
 /// Unstable. Used for the ongoing blueprint experimentations.
 struct PanelView (
-  "attr.rust.derive": "Copy"
+  "attr.rust.derive": "Copy",
+  "attr.rust.override_crate": "re_viewer"
 ) {
   is_expanded: bool (order: 100);
 }
+
```
Comment (teh-cmc) [inline crates/re_types/definitions/rerun/blueprint/panel_view.fbs:21]: x

**Comment by teh-cmc** [comment]
I have no clue why the C++ formatting check fails, I'm literally calling `just codegen`... I've rebased and everything, we'll see how that goes this time :shrug:

File: crates/re_types/src/datatypes/mesh_properties.rs [hunk 3]
```diff
@@ -94,88 +96,54 @@ impl ::re_types_core::Loggable for MeshProperties {
...
+                        ::re_types_core::external::arrow2::bitmap::Bitmap > = None; let
+                        offsets = ::re_types_core::external::arrow2::offset::Offsets:: <
+                        i32 > ::try_from_lengths(indices.iter().map(| opt | opt.as_ref()
+                        .map(| datum | datum.num_instances()).unwrap_or_default()))
+                        .unwrap().into(); ListArray::new(DataType::List(Box::new(Field {
+                        name : "item".to_owned(), data_type : DataType::UInt32,
+                        is_nullable : false, metadata : [].into(), })), offsets,
+                        PrimitiveArray::new(DataType::UInt32, indices_inner_data,
+                        indices_inner_bitmap,).boxed(), indices_bitmap,).boxed() } },
+                    ],
```
Comment (emilk) [inline crates/re_types/src/datatypes/mesh_properties.rs:126]: let's try to solve this problem, because this is 😬

**Review by emilk — COMMENTED** [review]
image

aren't these files marked as generated anymore? GitHub doesn't show the usual text as it does for other things:

File: crates/re_viewer/src/blueprint/panel_view.rs [hunk 4]
```diff
@@ -127,18 +131,18 @@ impl ::re_types_core::Loggable for PanelView {
...
     {
         re_tracing::profile_function!();
-        use ::arrow2::{array::*, buffer::*, datatypes::*};
         use ::re_types_core::{Loggable as _, ResultExt as _};
+        use arrow2::{array::*, buffer::*, datatypes::*};
         Ok({
             let arrow_data = arrow_data
                 .as_any()
-                .downcast_ref::<::arrow2::array::StructArray>()
+                .downcast_ref::<::re_types_core::external::arrow2::array::StructArray>()
```
Comment (emilk) [inline crates/re_viewer/src/blueprint/panel_view.rs:145]:
```suggestion
                .downcast_ref::<arrow2::array::StructArray>()
```

here and everywhere

File: crates/re_types_builder/src/codegen/rust/util.rs [hunk 5]
```diff
@@ -122,3 +122,16 @@ pub fn string_from_quoted(acc: &TokenStream) -> String {
...
+/// Returns the crate name of an object, accounting for overrides.
+pub fn crate_name(obj: &Object) -> String {
+    obj.try_get_attr::<String>(crate::ATTR_RUST_OVERRIDE_CRATE)
+        .unwrap_or_else(|| "re_types".to_owned())
+}
+
+/// Returns the module name of an object.
+//
+// NOTE: Might want a module override at some point.
+pub fn module_name(obj: &Object) -> String {
```
Comment (emilk) [inline crates/re_types_builder/src/codegen/rust/util.rs:135]: Let's make these members of the `impl Object` to make them easier to find

File: crates/re_types_builder/src/codegen/rust/serializer.rs [hunk 6]
```diff
@@ -40,7 +40,7 @@ pub fn quote_arrow_serializer(
 
     let quoted_bitmap = |var| {
         quote! {
-            let #var: Option<::arrow2::bitmap::Bitmap> = {
+            let #var: Option<::re_types_core::external::arrow2::bitmap::Bitmap> = {
```
Comment (emilk) [inline crates/re_types_builder/src/codegen/rust/serializer.rs:43]: Let's just `use ::re_types_core::external::arrow2;` at the top instead - this is way too hairy

Comment (teh-cmc) [reply crates/re_types_builder/src/codegen/rust/serializer.rs:43]: That's already the case actually, some search-n-replace seems to have been lost in translation :thinking:

File: crates/re_types/src/datatypes/mesh_properties.rs (hunk 3 above)
Comment (teh-cmc) [reply crates/re_types/src/datatypes/mesh_properties.rs]: there we go

**Comment by teh-cmc** [comment]
> image
> 
> aren't these files marked as generated anymore? GitHub doesn't show the usual text as it does for other things:

That's weird, the gitattributes [seem to be generated correctly](https://github.com/rerun-io/rerun/blob/cmc/arrow2convert_be_gone_3_for_real/crates/re_types/src/testing/components/.gitattributes)... I'll have a closer look once I get a minute.

File: rerun_cpp/src/rerun/blueprint/space_view_component.cpp [hunk 7]
```diff
@@ -0,0 +1,76 @@
...
+                    false
+                ),
+            });
+            return datatype;
+        }
+
+        Result<std::shared_ptr<arrow::StructBuilder>> SpaceViewComponent::new_arrow_array_builder(
+            arrow::MemoryPool *memory_pool
+        ) {
+            if (!memory_pool) {
```
Comment (emilk) [inline rerun_cpp/src/rerun/blueprint/space_view_component.cpp:25]: You are still on the old `.clang-format` file. I recently added `DerivePointerAlignment: false` to it that ensured pointer-stars get grouped with the type, so `arrow::MemoryPool* memory_pool`.

You should merge in `main` and run `just codegen` again before merging

**Review by emilk — APPROVED** [review]

**Pull request merged**